### DISCUSSION TITLE
Compare 2 exsiting files

### DISCUSSION THREAD

**Discussion by sf8321** [VerifyTests/Verify#799] [Show and tell]
Hello,
i have a commandline tool that converts a text file to an other text file and it is writen in c++. that tool i have rewriten in c# and i want 
write unit tests for that. In my Verify method i run the old tool to get the verified file. after that i run the new tool to create the received file. renamed both files to extension txt and call VerifyFile(reivedFile, settings). settings are UseFileName, UseDirectory. All looks good
i have 2 files af000215.verified.txt and af000215.received.txt. if i call VerifyFile my received file are deleted and get exception file not found. what is wrong here ? where is my mistake ?

**Comment by SimonCropp** [comment]
can you share the code to reproduce this issue?

**Comment by sf8321** [comment]
Hello,
yes.

    private static Task Verify(TestData testData)
    {
        
        //string verifiedFile = testData.RefFile.Replace(".ref.wup", ".verified.txt");
        //string receivedFile = testData.OutFile.Replace(".act.wup", ".txt");
        
        if (File.Exists(testData.RefFile) == false)
        {
            ExecuteOrgWupConverter(testData.WorkDir, testData.BaseDir, testData.InFile, testData.RefFile);
            //if (File.Exists(testData.RefFile))
            //{
            //    File.Delete(verifiedFile);
            //    File.Copy(testData.RefFile, verifiedFile);
            //}
        }

        ExecuteCSharpWupConverter(testData.WorkDir, testData.BaseDir, testData.InFile, testData.OutFile);

        var settings = new VerifySettings();
        settings.UseFileName(testData.FileName);
        settings.UseDirectory(DirectoryHelper.ToPlatformPath("Data/granitFlow_stark/data/test"));

        //if (File.Exists(testData.OutFile))
        //{
        //    File.Delete(receivedFile);
        //    File.Copy(testData.OutFile, receivedFile);
        //}

        //string text = File.ReadAllText(testData.OutFile);

        return Verifier.Verify(File.OpenRead(testData.OutFile),settings);
        //return Verifier.VerifyFile(receivedFile,settings);
    }

    [Theory]
    [MemberData(nameof(GetData))]
    public Task Test(TestData testData)
    {
        return Verify(testData);
    }

    public class TestData
    {
        public string FileName { get; set; } = null!;
        public string BaseDir { get; set; } = null!;
        public string DataPath { get; set; } = null!;
        public string InFile { get; set; } = null!;
        public string RefFile { get; set; } = null!;
        public string OutFile { get; set; } = null!;
        public string WorkDir { get; set; } = null!;
    }


    public static IEnumerable<object[]> GetData()
    {
        var baseDir = AppContext.BaseDirectory;
        var appRoot = baseDir.Substring(0,
            AppContext.BaseDirectory.LastIndexOf(Path.DirectorySeparatorChar + "bin", StringComparison.Ordinal));

        string dataBaseDir = DirectoryHelper.ToPlatformPath(DirectoryHelper.ToUnixPath(appRoot) + "/Data/granitFlow_stark");
        string dataPath = DirectoryHelper.ToPlatformPath(dataBaseDir + "/data/produnits");
        string testPath = DirectoryHelper.ToPlatformPath(dataBaseDir + "/data/test");
        string workDir = dataBaseDir;
        string inFile;
        string refFile;
        string outFile;


        if (Directory.Exists(testPath) == false)
        {
            Directory.CreateDirectory(testPath);
        }


        string path = DirectoryHelper.ToPlatformPath(dataPath);
        List<string> files = new List<string>();
        foreach (var file in Directory.GetFiles(path, "*"))
        {
            files.Add(Path.GetFileName(file));
        }

        foreach (var file in files)
        {
            inFile = DirectoryHelper.ToPlatformPath(dataPath + "/" + file);
            refFile = DirectoryHelper.ToPlatformPath(dataBaseDir + "/data/test/" + file + ".verified.wup");
            outFile = DirectoryHelper.ToPlatformPath(dataBaseDir + "/data/test/" + file + ".received.wup");
            yield return new object[]
            {
                new TestData
                {
                    FileName = file,
                    BaseDir = dataBaseDir,
                    DataPath = dataPath,
                    InFile = inFile,
                    RefFile = refFile,
                    OutFile = outFile,
                    WorkDir = workDir,
                }
            };
            break;
        }
    }

**Comment by sf8321** [comment]
Hello,
at the moment i have rewrite the test to use FileStream. but than i get this exception

System.IO.IOException: The process cannot access the file 'C:\workspace_csharp\WupConverter\WupConverter.Test\Data\granitFlow_stark\data\test\af000215...

System.IO.IOException
The process cannot access the file 'C:\workspace_csharp\WupConverter\WupConverter.Test\Data\granitFlow_stark\data\test\af000215.received.wup' because it is being used by another process.
   at System.IO.FileSystem.CopyFile(String sourceFullPath, String destFullPath, Boolean overwrite)
   at IoHelpers.TryCopyFileStream(String path, Stream stream) in /_/src/Verify/IoHelpers.cs:line 129
   at IoHelpers.WriteStream(String path, Stream stream) in /_/src/Verify/IoHelpers.cs:line 218
   at FileComparer.DoCompare(VerifySettings settings, FilePair file, Boolean previousTextFailed, Stream receivedStream) in /_/src/Verify/Compare/FileComparer.cs:line 26
   at VerifyEngine.GetResult(VerifySettings settings, FilePair file, Target target, Boolean previousTextFailed) in /_/src/Verify/Verifier/VerifyEngine.cs:line 42
   at VerifyEngine.HandleResults(List`1 targetList) in /_/src/Verify/Verifier/VerifyEngine.cs:line 52
   at VerifyTests.InnerVerifier.VerifyInner(Object root, Func`1 cleanup, IEnumerable`1 targets, Boolean doExpressionConversion) in /_/src/Verify/Verifier/InnerVerifier_Inner.cs:line 23
   at VerifyTests.InnerVerifier.VerifyStream(Stream stream, String extension, Object info) in /_/src/Verify/Verifier/InnerVerifier_Stream.cs:line 112
   at VerifyXunit.Verifier.<>c__DisplayClass7_0.<<Verify>b__0>d.MoveNext() in /_/src/Verify.Xunit/Verifier.cs:line 74
--- End of stack trace from previous location ---
   at Xunit.Sdk.TestInvoker`1.<>c__DisplayClass48_0.<<InvokeTestMethodAsync>b__1>d.MoveNext() in /_/src/xunit.execution/Sdk/Frameworks/Runners/TestInvoker.cs:line 264
--- End of stack trace from previous location ---
   at Xunit.Sdk.ExecutionTimer.AggregateAsync(Func`1 asyncAction) in /_/src/xunit.execution/Sdk/Frameworks/ExecutionTimer.cs:line 48
   at Xunit.Sdk.ExceptionAggregator.RunAsync(Func`1 code) in /_/src/xunit.core/Sdk/ExceptionAggregator.cs:line 90

**Comment by sf8321** [comment]
Hello,
i make a simpler testcase xunit/verifytest

    private static Task VerifyTest()
    {
        
        var baseDir = AppContext.BaseDirectory;
        var appRoot = baseDir.Substring(0,
            AppContext.BaseDirectory.LastIndexOf(Path.DirectorySeparatorChar + "bin", StringComparison.Ordinal));

        string verifiedFile = Path.Combine(appRoot, "StarkUnitTest1.Test2.verified.txt");
        string receivedFile = Path.Combine(appRoot, "StarkUnitTest1.Test2.received.txt");
        
        if (File.Exists(verifiedFile) == false)
        {
            string txt = "Hallo";
            File.WriteAllText(verifiedFile,txt);
        }

        string txt2 = "Hallole";
        
        File.WriteAllText(receivedFile,txt2);
        
        return Verifier.VerifyFile(receivedFile);
    }

    
    
    [Fact]
    public Task Test2()
    {
        return VerifyTest();
    }

**Comment by sf8321** [comment]
Hello, 
i think i found my problem.
the second file must not have the name ....received.txt any other name are good. The VerififyFile method creates the ....received.txt file and after that it will compare the 2 files. 

Will it not better to have same VerifyExistingFile(...). It is only a suggestion. yet i have no problem anymore but i need same time to find out how it works.


Changed example

    private static Task VerifyTest()
    {
        
        var baseDir = AppContext.BaseDirectory;
        var appRoot = baseDir.Substring(0,
            AppContext.BaseDirectory.LastIndexOf(Path.DirectorySeparatorChar + "bin", StringComparison.Ordinal));

        string verifiedFile = Path.Combine(appRoot, "StarkUnitTest1.Test2.verified.txt");
        string receivedFile = Path.Combine(appRoot, "StarkUnitTest1.Test2.test.txt");
        
        if (File.Exists(verifiedFile) == false)
        {
            string txt = "Hallo";
            File.WriteAllText(verifiedFile,txt);
        }

        string txt2 = "Hallole";
        
        File.WriteAllText(receivedFile,txt2);
        
        return Verifier.VerifyFile(receivedFile);
    }

    
    
    [Fact]
    public Task Test2()
    {
        return VerifyTest();
    }

**Reply by SimonCropp** [reply]
what code are you actually testing?

**Reply by SimonCropp** [reply]
the verified and received files are controlled by Verify. you should not programmatically manipulate them.

**Comment by sf8321** [comment]
Hello,
ok my fault. i have solved this in an other way. without VerifyTest. thk.
